### PR TITLE
Add Bluefin Payment Processor 

### PR DESCRIPTION
**Pull request by tehreem-sadat**

### CODE REVIEW

File: ecommerce/extensions/payment/processors/bluefin.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    
+    def __init__(self, site):
+        """
+        Constructs a new instance of the Bluefin processor.
+
+        Raises:
+            KeyError: If no settings configured for this payment processor.
+        """
+        super(Bluefin, self).__init__(site)
+        configuration = self.configuration
```
Comment (asamolion) [inline ecommerce/extensions/payment/processors/bluefin.py]: Where is the bluefin configuration saved?

File: ecommerce/extensions/payment/processors/bluefin.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        self.account_id =  configuration['merchant_account_id']
+        self.api_key = configuration['api_access_key']
+        self.post_url = configuration['post_api_url']
+        
+    def get_transaction_parameters(self, basket, request=None, use_client_side_checkout=True, **kwargs):
```
Comment (asamolion) [inline ecommerce/extensions/payment/processors/bluefin.py]: What is the purpose of this function?

File: ecommerce/extensions/payment/processors/bluefin.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        raise NotImplementedError('The Bluefin payment processor does not support transaction parameters.')
+
+    def _get_basket_amount(self, basket):
+        return str((basket.total_incl_tax * 100).to_integral_value())
```
Comment (asamolion) [inline ecommerce/extensions/payment/processors/bluefin.py]: Shouldn't there be a configuration of currency type here as well?

File: ecommerce/extensions/payment/processors/bluefin.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        request_data['street_address1'] = request_data.pop('address_line1', None)
+        request_data['street_address2'] = request_data.pop('address_line2', None)
+        request_data['state'] = request_data.pop('country', None)
+        request_data['zip'] = request_data.pop('postal_code', None)
+        request_data['eToken'] = request_data.pop('bluefin_token', None)
+
+        request_data.pop('csrfmiddlewaretoken', None)
+        request_data.pop('basket', None)
+        
+        request_data['account_id'] = self.account_id
```
Comment (asamolion) [inline ecommerce/extensions/payment/processors/bluefin.py]: if `request_data` is a dict then use `.get()` to get the values wherever possible. Because `.get()` has a failsafe in case the value is not present on the dict.

File: ecommerce/extensions/payment/processors/bluefin.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    
+        return HandledProcessorResponse(
+            transaction_id=transaction_id,
+            total=total,
+            currency=currency,
+            card_number=response['last4'],
+            card_type=response['card_brand']
+        )
+    
+    def issue_credit(self, order_number, basket, reference_number, amount, currency):
```
Comment (asamolion) [inline ecommerce/extensions/payment/processors/bluefin.py]: What is the purpose of this function?

File: ecommerce/extensions/payment/views/bluefin.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    """ Bluefin payment handler.
+
+    The payment form should POST here. This view will handle Bluefin processing 
+    and redirection of the user to the receipt page.
+    """
+
+    form_class = BluefinSubmitForm
+
+    @property
+    def payment_processor(self):
```
Comment (asamolion) [inline ecommerce/extensions/payment/views/bluefin.py]: Why is this necessary?

File: ecommerce/extensions/payment/views/bluefin.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        form_data = form.cleaned_data
+        basket = form_data['basket']
+        token = self.request.POST['bluefin_token']
+        order_number = basket.order_number
+
+        basket_add_organization_attribute(basket, self.request.POST)
+
+        try:
+           billing_address = BillingAddress (
+               first_name=self.request.POST['first_name'],
```
Comment (asamolion) [inline ecommerce/extensions/payment/views/bluefin.py]: use `.get()` to get the values.

File: ecommerce/settings/_oscar.py [hunk 8]
```diff
@@ -136,6 +137,12 @@
...
         },
+        'bluefin': {
```
Comment (asamolion) [inline ecommerce/settings/_oscar.py]: Is this the correct syntax?

File: ecommerce/settings/_oscar.py [hunk 9]
```diff
@@ -136,6 +137,12 @@
...
+            'merchant_account_id': '220614981961',
+            'api_access_key': '114be6d774963c3fbba2a66d08529ebd',
+            'post_api_url': 'https://cert.payconex.net/api/qsapi/3.8',
+
```
Comment (asamolion) [inline ecommerce/settings/_oscar.py]: Please remove whitespace. Also check PEP8 on the trailing comma.

File: ecommerce/settings/devstack.py [hunk 10]
```diff
@@ -46,6 +46,12 @@
...
         },
+        'bluefin': {
+            'merchant_account_id': '220614981961',
```
Comment (asamolion) [inline ecommerce/settings/devstack.py]: Shouldn't these values be coming from a config file?

File: ecommerce/static/js/payment_processors/bluefin.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
+/* Declare iFrame variable for setting in function later */
+var paymentiFrame;
+
+/* parameter definition for iFrame */
+var params = {
+    create    : true,
+    iframeId  : "payment_iframe",
+    settings  : {
+        account         : "220614981961",
```
Comment (asamolion) [inline ecommerce/static/js/payment_processors/bluefin.js]: No need to evenly space the colons. Read the Open edX JavaScript style guide and code accordingly.

File: ecommerce/static/js/payment_processors/bluefin.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+function submitPayment(post_url){
+    paymentiFrame.encrypt()
+    .failure( function (err) {
+        //Show iframe failure response on page.
+        document.getElementById('iframe-response').innerHTML = JSON.stringify(err, null, 2);
+    })
+    .invalidInput( function (data) {
+        // Show iframe error message on page.
+        error = data.invalidInputs[0]
+        error_string = `Invalid card details.\n ${error.field} field: ${error.message}`
```
Comment (asamolion) [inline ecommerce/static/js/payment_processors/bluefin.js]: Don't use template literals because they are not supported in many browsers.

File: ecommerce/static/js/payment_processors/bluefin.js [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        
+        document.getElementById('iframe-response').innerHTML = `
```
Comment (asamolion) [inline ecommerce/static/js/payment_processors/bluefin.js]: again, don't use template literals.

File: ecommerce/static/js/payment_processors/bluefin.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                        <div id="iframe" style="overflow:hidden;"></div>
+                    </fieldset>
+                    <div id="iframe-response"></div>
+                </div>
+            `
+            $( bluefin_iframe_code ).insertAfter( ".pci-fields" );
+            let card_div  = document.getElementsByClassName("pci-fields");
+            card_div[0].style.display = "none"
+
+            //dummy data
```
Comment (asamolion) [inline ecommerce/static/js/payment_processors/bluefin.js]: Does this dummy data need to be pushed to production? If not, then why is this hard coded?

File: ecommerce/settings/devstack.py (hunk 10 above)
Comment (tehreem-sadat) [reply ecommerce/settings/devstack.py]: I know that it shouldn't be here but there is no auth.json file for ecommerce and other processor's secret keys have also been placed here. 
Where should i place this other than this file ?

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 1 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: _oscar.py and devstack.py

File: ecommerce/settings/_oscar.py (hunk 8 above)
Comment (tehreem-sadat) [reply ecommerce/settings/_oscar.py]: yes as it is dict object data saved in PAYMENT_PROCESSOR_CONFIG variab
le.

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 14 above)
Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: Yes as there is by default full payment form (containing card details and person info) and we are using iframe instead of it's card fields so we have to set some dummy values to it otherwise it would fail in form validation.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 5 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: Used for refund transaction. But as it seems that there is no Refund policy in UCSD so it is not compulsory to give it's implementation.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 6 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/views/bluefin.py]: As there are multiple processors (cybersource, stripe)  to handle credit card transaction and any one can be used as back-end.so we have to make sure that our request should be sent to our submit view whenever our processor is activated at the back-end.
And to make processor activated (that which processor should be used), we change some settings at the admin site of lms.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 2 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: It generate a dictionary of extra parameters requires to complete a transaction.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 4 above)
Comment (mumarkhan999) [reply ecommerce/extensions/payment/processors/bluefin.py]: done.

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 13 above)
Comment (mumarkhan999) [reply ecommerce/static/js/payment_processors/bluefin.js]: done

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 12 above)
Comment (mumarkhan999) [reply ecommerce/static/js/payment_processors/bluefin.js]: done.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 7 above)
Comment (mumarkhan999) [reply ecommerce/extensions/payment/views/bluefin.py]: done

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 4 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: done.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 7 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/views/bluefin.py]: done.

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 12 above)
Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: done

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 13 above)
Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: done

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 11 above)
Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: done

**Comment by tehreem-sadat** [comment]
@asamolion done with requested changes.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 1 above)
Comment (asamolion) [reply ecommerce/extensions/payment/processors/bluefin.py]: Alright.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 4 above)
Comment (asamolion) [reply ecommerce/extensions/payment/processors/bluefin.py]: Please use `.get()` not `.pop()`. Since `.pop()` removes the value from the dict.

Comment (asamolion) [reply ecommerce/extensions/payment/processors/bluefin.py]: If it is necessary to remove the value then you can use it.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 5 above)
Comment (asamolion) [reply ecommerce/extensions/payment/processors/bluefin.py]: Okay, then I believe you should raise a `NotImplemented` Exception.

File: ecommerce/settings/_oscar.py [hunk 15]
```diff
@@ -136,6 +137,12 @@
...
         },
+        'bluefin': {
+            'merchant_account_id': '220614981961',
```
Comment (asamolion) [inline ecommerce/settings/_oscar.py]: This is our sandbox configuration so it should be moved in `ecommerce/settings/devstack.py` file. The `production.py` file overrides these settings with proper settings from the environment.

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 14 above)
Comment (asamolion) [reply ecommerce/static/js/payment_processors/bluefin.js]: But, if we push this to production, will these values show up when a user goes to the payment form?

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 4 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: basically i want to rename key form 'address_line1' to 'street_address1'. So using pop i can do it in single line which is i guess better than two lines code of using get and then using pop.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 5 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: okay (Y)

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 14 above)
Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: style.display = "none" ( line 111 )  is hiding it. So it would not be displayed to the user, our iframe will take it's place and we will send our iframe fields after validation to the back-end.
Moreover this js file will only run in case of Bluefin processor so in case of other processors (cybersource, stripe) normal flow of Payment Form without iframe  and without any dummy values will run.

File: ecommerce/settings/_oscar.py (hunk 9 above)
Comment (tehreem-sadat) [reply ecommerce/settings/_oscar.py]: done.

File: ecommerce/extensions/payment/processors/bluefin.py (hunk 1 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/processors/bluefin.py]: In edx oscar.py contain processor configuration keys with none values (ie: account_id: None, api_key=None) and devstack.py contain configuration keys with actual value (actual account id and api access key). So I have updated these change

File: ecommerce/settings/devstack.py (hunk 10 above)
Comment (tehreem-sadat) [reply ecommerce/settings/devstack.py]: In edx oscar.py contain processor configuration keys with none values (ie: account_id: None, api_key=None) and devstack.py contain configuration keys with actual value (actual account id and api access key). So I have updated these changes.

File: ecommerce/settings/_oscar.py [hunk 16]
```diff
@@ -135,7 +136,12 @@
...
-        },
+        }
+        'bluefin': {
+            'merchant_account_id': None,
+            'api_access_key': None,
+            'post_api_url': 'https://cert.payconex.net/api/qsapi/3.8',
```
Comment (asamolion) [inline ecommerce/settings/_oscar.py]: Shouldn't this key be moved to `devstack.py` as well?

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 14 above)
Comment (asamolion) [reply ecommerce/static/js/payment_processors/bluefin.js]: So shouldn't we leave these values empty or use better defaults?

Comment (tehreem-sadat) [reply ecommerce/static/js/payment_processors/bluefin.js]: we have to give valid dummy values to these fields. What do you mean by better defaults ?

File: ecommerce/settings/_oscar.py (hunk 16 above)
Comment (tehreem-sadat) [reply ecommerce/settings/_oscar.py]: done.

File: ecommerce/extensions/payment/forms.py [hunk 17]
```diff
@@ -229,3 +229,33 @@ def clean_basket(self):
             Applicator().apply(basket, self.request.user, self.request)
 
         return basket
+
+
+class BluefinSubmitForm(forms.Form):
```
Comment (waheedahmed) [inline ecommerce/extensions/payment/forms.py]: Any specific reason for not using already defined `PaymentForm`? Because I don't see any difference in both, also you can get leverage from the other fields of that form.

File: ecommerce/extensions/payment/views/bluefin.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        form_data = form.cleaned_data
+        basket = form_data['basket']
+        token = self.request.POST['bluefin_token']
+        order_number = basket.order_number
+
+        basket_add_organization_attribute(basket, self.request.POST)
+
+        try:
+            billing_address = BillingAddress(
+                first_name=self.request.POST.get('first_name'),
```
Comment (waheedahmed) [inline ecommerce/extensions/payment/views/bluefin.py]: You can user `form_data` to get cleaned values here if you use `PaymentForm`

File: ecommerce/extensions/payment/views/bluefin.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    form_class = BluefinSubmitForm
+
+    @property
+    def payment_processor(self):
+        return Bluefin(self.request.site)
+
+    def form_valid(self, form):
+        form_data = form.cleaned_data
+        basket = form_data['basket']
+        token = self.request.POST['bluefin_token']
```
Comment (waheedahmed) [inline ecommerce/extensions/payment/views/bluefin.py]: Define it in the form.

File: ecommerce/extensions/payment/views/bluefin.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+            logger.exception(
+                'An error occurred while parsing the billing address for \
+                    basket [%d]. No billing address will be stored for \
+                        the resulting order [%s].',
+                basket.id,
+                order_number)
+            billing_address = None
+
+        try:
+            self.handle_payment(self.request.POST.dict(), basket)
```
Comment (waheedahmed) [inline ecommerce/extensions/payment/views/bluefin.py]: Pass form's cleaned data instead of POST data if you user `PaymentForm`.

File: ecommerce/static/js/payment_processors/bluefin.js (hunk 14 above)
Comment (waheedahmed) [reply ecommerce/static/js/payment_processors/bluefin.js]: I also don't get get why this dummy data is required here?

File: ecommerce/extensions/payment/forms.py (hunk 17 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/forms.py]: Have made these changes in my last commit.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 18 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/views/bluefin.py]: Have made these changes in my last commit.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 19 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/views/bluefin.py]: Have made these changes in my last commit.

File: ecommerce/extensions/payment/views/bluefin.py (hunk 20 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/views/bluefin.py]: Have made these changes in my last commit.

File: ecommerce/extensions/payment/forms.py (hunk 17 above)
Comment (tehreem-sadat) [reply ecommerce/extensions/payment/forms.py]: Have made these changes in my last commit.

**Comment by tehreem-sadat** [comment]
@waheedahmed  I have made all requested changes and have also removed dummy values. Kindly review it again.

File: ecommerce/extensions/payment/forms.py [hunk 21]
```diff
@@ -229,3 +229,12 @@ def clean_basket(self):
...
         return basket
+
+
+class BluefinSubmitForm(PaymentForm):
+    """
+    Form for Bluefin token submissions.
+    """
+
+    bluefin_token = forms.CharField(widget=forms.HiddenInput(), required=True)
+
```
Comment (waheedahmed) [inline ecommerce/extensions/payment/forms.py]: Add a new line.

Comment (tehreem-sadat) [reply ecommerce/extensions/payment/forms.py]: done.
